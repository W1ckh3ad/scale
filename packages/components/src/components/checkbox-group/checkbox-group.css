.checkbox-group {
  display: inline-flex;
  flex-direction: column;
}

.checkbox-group__container {
  display: flex;
  flex-direction: column;
  margin-left: 1.5rem;
}

.checkbox-group__checkbox {
  margin-top: 0.5rem;
}

::slotted(*) {
  padding-top: 5px;
}
