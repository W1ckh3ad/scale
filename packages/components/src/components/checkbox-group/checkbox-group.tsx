/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { Component, h, Host, Listen, Element, State } from '@stencil/core';
import { CheckboxState } from './utils/utils';

@Component({
  tag: 'scale-checkbox-group',
  styleUrl: './checkbox-group.css',
  shadow: true,
})
export class CheckboxGroup {
  initialLoad: boolean = true;
  statusChanged: boolean = false;
  @State() groupStatus: CheckboxState[] = [];
  @Element() hostElement: HTMLElement;
  @Listen('scaleChange')
  scaleChangeHandler() {
    this.setGroupStatusState();
  }

  componentDidLoad() {
    this.setGroupStatusState();
  }

  setGroupStatusState() {
    const checkboxes = Array.from(
      this.hostElement.querySelectorAll('scale-checkbox')
    );
    const newState: CheckboxState[] = [];
    for (let i = 0; i < checkboxes.length; i++) {
      newState[i] = {
        id: checkboxes[i].inputId,
        checked: checkboxes[i].checked,
        disabled: checkboxes[i].disabled ? checkboxes[i].disabled : false,
        indeterminate: checkboxes[i].indeterminate,
      };
    }

    if (this.initialLoad) {
      this.handleMasterCheckBoxState(newState);
      this.initialLoad = false;
      this.statusChanged = true;
    } else if (this.initialLoad == false) {
      this.handleMasterCheckBoxState(newState);
      this.statusChanged = true;
    }
    this.groupStatus = newState;
  }

  handleMasterCheckBoxState(newState: CheckboxState[]) {
    let checkedCounter = 0;
    for (let i = 1; i < newState.length; i++) {
      if (newState[i].checked) {
        checkedCounter += 1;
      }
    }
    let masterCheckbox;
    masterCheckbox = this.hostElement.querySelectorAll('scale-checkbox');
    //indeterminate
    if (checkedCounter < newState.length - 1 && checkedCounter != 0) {
      masterCheckbox[0].setAttribute('indeterminate', 'true');
      masterCheckbox[0].removeAttribute('checked');
      this.setMasterChanges();
    }
    //empty
    else if (checkedCounter == 0) {
      masterCheckbox[0].removeAttribute('indeterminate');
      masterCheckbox[0].removeAttribute('checked');
      this.setMasterChanges();
    }
    //checked
    else if (checkedCounter == newState.length - 1) {
      masterCheckbox[0].setAttribute('checked', 'true');
      masterCheckbox[0].removeAttribute('indeterminate');
      this.setMasterChanges();
    }
  }

  setMasterChanges() {
    const checkboxes = Array.from(
      this.hostElement.querySelectorAll('scale-checkbox')
    );
    const newState: CheckboxState[] = [];
    for (let i = 0; i < checkboxes.length; i++) {
      newState[i] = {
        id: checkboxes[i].inputId,
        checked: checkboxes[i].checked,
        disabled: checkboxes[i].disabled ? checkboxes[i].disabled : false,
        indeterminate: checkboxes[i].indeterminate,
      };
    }
    this.groupStatus = newState;
    for (let i = 0; i < newState.length; i++) {
      console.log(newState[i]);
    }
  }

  render() {
    return (
      <Host>
        <div class="checkbox-group">
          <div class="checkbox-group__label">
            <slot name="checkbox-header" />
          </div>
          <div class="checkbox-group__container">
            <slot name="checkbox-list" />
          </div>
        </div>
      </Host>
    );
  }
}
