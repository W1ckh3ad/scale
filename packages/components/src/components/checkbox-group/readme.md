# scale-checkbox-group



<!-- Auto Generated Below -->


## Dependencies

### Depends on

- [scale-checkbox](../checkbox)

### Graph
```mermaid
graph TD;
  scale-checkbox-group --> scale-checkbox
  scale-checkbox --> scale-icon-action-success
  scale-checkbox --> scale-icon-action-play
  style scale-checkbox-group fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
