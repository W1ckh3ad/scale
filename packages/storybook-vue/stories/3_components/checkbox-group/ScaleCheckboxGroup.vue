<template>
  <scale-checkbox-group>
    <slot></slot>
    <slot></slot>
  </scale-checkbox-group>
</template>

<script>
import { action } from '@storybook/addon-actions';

export default {
  props: {},
  methods: {},
};
</script>
