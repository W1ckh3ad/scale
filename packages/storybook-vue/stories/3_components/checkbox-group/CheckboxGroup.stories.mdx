import {
  Meta,
  Story,
  ArgsTable,
  Canvas,
  Description,
} from '@storybook/addon-docs/blocks';
import ScaleCheckboxGroup from './ScaleCheckboxGroup.vue';
import { action } from '@storybook/addon-actions';

<Meta
  title="Beta Components/Checkbox Group"
  component={ScaleCheckboxGroup}
  argTypes={{}}
/>

export const StandardTemplate = (args, { argTypes }) => ({
  props: {
    width: { type: String },
    ...ScaleCheckboxGroup.props,
  },
  computed: {
    wrapperStyles() {
      return {
        maxWidth: this.width,
      };
    },
  },
  template: `
    <div :style="wrapperStyles">
    <scale-checkbox-group>
  <div slot="checkbox-header">
    <scale-checkbox
      input-id="header-checkbox-0"
      value="1"
      label="checkbox"
      name="nameOfCheckbox"
      checked
      helper-text="helperText"
    >
    </scale-checkbox>
  </div>
  <div slot="checkbox-list">
    <div style="margin-top: 0.5rem">
      <scale-checkbox
        value="2"
        label="checkbox"
        name="nameOfCheckbox"
      ></scale-checkbox>
    </div>
    <div style="margin-top: 0.5rem">
      <scale-checkbox value="3" name="nameOfCheckbox" checked
        >Checkbox Slot</scale-checkbox
      >
    </div>
  </div>
</scale-checkbox-group>
    </div>
  `,
});

export const CheckboxDisableTemplate = (args, { argTypes }) => ({
  props: {
    width: { type: String },
    ...ScaleCheckboxGroup.props,
  },
  computed: {
    wrapperStyles() {
      return {
        maxWidth: this.width,
      };
    },
  },
  template: `
    <div :style="wrapperStyles">
    <scale-checkbox-group>
  <div slot="checkbox-header">
    <scale-checkbox
      input-id="header-checkbox-1"
      value="1"
      label="checkbox"
      name="nameOfCheckbox"
      checked
      helper-text="helperText"
    >
    </scale-checkbox>
  </div>
  <div slot="checkbox-list">
    <div style="margin-top: 0.5rem">
      <scale-checkbox
        value="2"
        label="checkbox"
        name="nameOfCheckbox"
      ></scale-checkbox>
    </div>
     <div style="margin-top: 0.5rem">
      <scale-checkbox
        value="3"
        label="checkbox"
        name="nameOfCheckbox"
        disabled
      ></scale-checkbox>
    </div>
     <div style="margin-top: 0.5rem">
      <scale-checkbox
        value="4"
        label="checkbox"
        name="nameOfCheckbox"
      ></scale-checkbox>
    </div>
       <div style="margin-top: 0.5rem">
      <scale-checkbox
        value="5"
        label="checkbox"
        name="nameOfCheckbox"
        checked
        disabled
      ></scale-checkbox>
    </div>
  </div>
</scale-checkbox-group>
    </div>
  `,
});

export const DisabledTemplate = (args, { argTypes }) => ({
  props: {
    width: { type: String },
    ...ScaleCheckboxGroup.props,
  },
  computed: {
    wrapperStyles() {
      return {
        maxWidth: this.width,
      };
    },
  },
  template: `
    <div :style="wrapperStyles">
    <scale-checkbox-group>
  <div slot="checkbox-header">
    <scale-checkbox
      input-id="header-checkbox-2"
      value="1"
      label="checkbox"
      name="nameOfCheckbox"
      checked
      helper-text="helperText"
      disabled
    >
    </scale-checkbox>
  </div>
  <div slot="checkbox-list">
    <div style="margin-top: 0.5rem">
      <scale-checkbox
        value="2"
        label="checkbox"
        name="nameOfCheckbox"
      ></scale-checkbox>
    </div>
  </div>
</scale-checkbox-group>
    </div>
  `,
});

export const ErrorTemplate = (args, { argTypes }) => ({
  props: {
    width: { type: String },
    ...ScaleCheckboxGroup.props,
  },
  computed: {
    wrapperStyles() {
      return {
        maxWidth: this.width,
      };
    },
  },
  template: `
    <div :style="wrapperStyles">
    <scale-checkbox-group>
  <div slot="checkbox-header">
    <scale-checkbox status="error"
      input-id="header-checkbox-2"
      helper-text="Error helperText"
      value="1"
      label="checkbox"
      name="nameOfCheckbox"
      checked
      helper-text="helperText"
    >
    </scale-checkbox>
  </div>
  <div slot="checkbox-list">
    <div style="margin-top: 0.5rem">
      <scale-checkbox
        disabled
        value="2"
        label="checkbox"
        name="nameOfCheckbox"
      ></scale-checkbox>
    </div>
  </div>
</scale-checkbox-group>
    </div>
  `,
});

## Standard

<Canvas withSource="none">
  <Story
    name="Standard"
    args={{
      label: 'Checkbox Group',
    }}
  >
    {StandardTemplate.bind({})}
  </Story>
</Canvas>

<ArgsTable story="Standard" />

```html
<scale-checkbox-group>
  <div slot="checkbox-header">
    <scale-checkbox
      input-id="header-checkbox-0"
      value="1"
      label="checkbox"
      name="nameOfCheckbox"
      checked
      helper-text="helperText"
    >
    </scale-checkbox>
  </div>
  <div slot="checkbox-list">
    <div style="margin-top: 0.5rem">
      <scale-checkbox
        value="2"
        label="checkbox"
        name="nameOfCheckbox"
      ></scale-checkbox>
    </div>
    <div style="margin-top: 0.5rem">
      <scale-checkbox value="3" name="nameOfCheckbox" checked
        >Checkbox Slot</scale-checkbox
      >
    </div>
  </div>
</scale-checkbox-group>
```

## Error

<Canvas withSource="none">
  <Story
    name="Error"
    args={{
      label: 'Checkbox Group',
    }}
  >
    {ErrorTemplate.bind({})}
  </Story>
</Canvas>

```html
<scale-checkbox-group>
  <div slot="checkbox-header">
    <scale-checkbox
      status="error"
      input-id="header-checkbox-0"
      value="1"
      label="checkbox"
      name="nameOfCheckbox"
      checked
      helper-text="helperText"
    >
    </scale-checkbox>
  </div>
  <div slot="checkbox-list">
    <div style="margin-top: 0.5rem">
      <scale-checkbox
        value="2"
        label="checkbox"
        name="nameOfCheckbox"
      ></scale-checkbox>
    </div>
    <div style="margin-top: 0.5rem">
      <scale-checkbox value="3" name="nameOfCheckbox" checked
        >Checkbox Slot</scale-checkbox
      >
    </div>
  </div>
</scale-checkbox-group>
```

## Checkbox Disabled

<Canvas withSource="none">
  <Story
    name="Checkbox Disabled"
    args={{
      label: 'Checkbox Group',
    }}
  >
    {CheckboxDisableTemplate.bind({})}
  </Story>
</Canvas>

```html
<scale-checkbox-group>
  <div slot="checkbox-header">
    <scale-checkbox
      input-id="header-checkbox-1"
      value="1"
      label="checkbox"
      name="nameOfCheckbox"
      checked
      helper-text="helperText"
    >
    </scale-checkbox>
  </div>
  <div slot="checkbox-list">
    <div style="margin-top: 0.5rem">
      <scale-checkbox
        value="2"
        label="checkbox"
        name="nameOfCheckbox"
      ></scale-checkbox>
    </div>
    <div style="margin-top: 0.5rem">
      <scale-checkbox
        value="3"
        label="checkbox"
        name="nameOfCheckbox"
        disabled
      ></scale-checkbox>
    </div>
    <div style="margin-top: 0.5rem">
      <scale-checkbox
        value="4"
        label="checkbox"
        name="nameOfCheckbox"
      ></scale-checkbox>
    </div>
    <div style="margin-top: 0.5rem">
      <scale-checkbox
        value="5"
        label="checkbox"
        name="nameOfCheckbox"
        checked
        disabled
      ></scale-checkbox>
    </div>
  </div>
</scale-checkbox-group>
```

## Checkbox Group Disabled

<Canvas withSource="none">
  <Story
    name="Checkbox-Group Disabled"
    args={{
      label: 'Checkbox Group',
    }}
  >
    {DisabledTemplate.bind({})}
  </Story>
</Canvas>

```html
<scale-checkbox-group>
  <div slot="checkbox-header">
    <scale-checkbox
      input-id="header-checkbox-2"
      value="1"
      label="checkbox"
      name="nameOfCheckbox"
      checked
      helper-text="helperText"
      disabled
    >
    </scale-checkbox>
  </div>
  <div slot="checkbox-list">
    <div style="margin-top: 0.5rem">
      <scale-checkbox
        value="2"
        label="checkbox"
        name="nameOfCheckbox"
      ></scale-checkbox>
    </div>
  </div>
</scale-checkbox-group>
```
