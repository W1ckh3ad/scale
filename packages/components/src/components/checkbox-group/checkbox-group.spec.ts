import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { CheckboxGroup } from './checkbox-group';

describe('CheckboxGroup', () => {
  let page: SpecPage;

  it('should match standard snapshot', async () => {
    page = await newSpecPage({
      components: [CheckboxGroup],
      html: ` <scale-checkbox-group>
      <div slot="checkbox-header">
        <scale-checkbox
          input-id="header-checkbox"
          value="1"
          label="checkbox"
          name="nameOfCheckbox"
          helper-text="helperText"
        >
        </scale-checkbox>
      </div>
      <div slot="checkbox-list">
        <div style="margin-top: 0.5rem">
          <scale-checkbox
            value="2"
            label="checkbox"
            name="nameOfCheckbox"
          ></scale-checkbox>
        </div>
        <div style="margin-top: 0.5rem">
          <scale-checkbox value="3" name="nameOfCheckbox" checked
            >Checkbox Slot</scale-checkbox
          >
        </div>
        <div style="margin-top: 0.5rem">
          <scale-checkbox
            value="4"
            label="checkbox"
            name="nameOfCheckbox"
            disabled
          ></scale-checkbox>
        </div>
        <div style="margin-top: 0.5rem">
          <scale-checkbox
            value="5"
            label="checkbox"
            name="nameOfCheckbox"
            checked
          >
          </scale-checkbox>
        </div>
        <div style="margin-top: 0.5rem">
          <scale-checkbox
            value="6"
            label="checkbox"
            name="nameOfCheckbox"
            checked
          >
          </scale-checkbox>
        </div>
        <div style="margin-top: 0.5rem">
          <scale-checkbox
            value="7"
            label="checkbox"
            name="nameOfCheckbox"
            checked
          >
          </scale-checkbox>
        </div>
        <div style="margin-top: 0.5rem">
          <scale-checkbox
            value="8"
            label="checkbox"
            name="nameOfCheckbox"
            checked
          >
          </scale-checkbox>
        </div>
        <div style="margin-top: 0.5rem">
          <scale-checkbox
            value="9"
            label="checkbox"
            name="nameOfCheckbox"
            disabled
            checked
          >
          </scale-checkbox>
        </div>
      </div>
    </scale-checkbox-group>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
