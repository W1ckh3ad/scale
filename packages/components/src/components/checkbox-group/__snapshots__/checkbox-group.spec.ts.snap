// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxGroup should match standard snapshot 1`] = `
<scale-checkbox-group>
  <mock:shadow-root>
    <div class="checkbox-group">
      <div class="checkbox-group__label">
        <slot name="checkbox-header"></slot>
      </div>
      <div class="checkbox-group__container">
        <slot name="checkbox-list"></slot>
      </div>
    </div>
  </mock:shadow-root>
  <div slot="checkbox-header">
    <scale-checkbox helper-text="helperText" input-id="header-checkbox" label="checkbox" name="nameOfCheckbox" value="1"></scale-checkbox>
  </div>
  <div slot="checkbox-list">
    <div style="margin-top: 0.5rem;">
      <scale-checkbox label="checkbox" name="nameOfCheckbox" value="2"></scale-checkbox>
    </div>
    <div style="margin-top: 0.5rem;">
      <scale-checkbox checked="" name="nameOfCheckbox" value="3">
        Checkbox Slot
      </scale-checkbox>
    </div>
    <div style="margin-top: 0.5rem;">
      <scale-checkbox disabled="" label="checkbox" name="nameOfCheckbox" value="4"></scale-checkbox>
    </div>
    <div style="margin-top: 0.5rem;">
      <scale-checkbox checked="" label="checkbox" name="nameOfCheckbox" value="5"></scale-checkbox>
    </div>
    <div style="margin-top: 0.5rem;">
      <scale-checkbox checked="" label="checkbox" name="nameOfCheckbox" value="6"></scale-checkbox>
    </div>
    <div style="margin-top: 0.5rem;">
      <scale-checkbox checked="" label="checkbox" name="nameOfCheckbox" value="7"></scale-checkbox>
    </div>
    <div style="margin-top: 0.5rem;">
      <scale-checkbox checked="" label="checkbox" name="nameOfCheckbox" value="8"></scale-checkbox>
    </div>
    <div style="margin-top: 0.5rem;">
      <scale-checkbox checked="" disabled="" label="checkbox" name="nameOfCheckbox" value="9"></scale-checkbox>
    </div>
  </div>
</scale-checkbox-group>
`;
